### PR TITLE
Reader: Add following feed Import/Export OPML buttons

### LINKED ISSUE #943: Reader: Export feeds via OPML
We should allow OPML exports for feed subscriptions.

Raised by @blowery

### LINKED ISSUE #944: Reader: Import feeds via OPML
We should allow OPML imports for feed subscriptions.

Raised by @blowery

### PR DESCRIPTION
**Pull request by TooTallNate**
Implements @jancavan's new reader following edit page design as prescribed in https://github.com/Automattic/wp-calypso/pull/3878#issuecomment-194991987. Also adds the new OPML Import/Export buttons in the process.

@blowery @bluefuton @jancavan I haven't yet figured out the `z-index()` SCSS function, which we'll now need for the `SubscribeFormResults` sub-component since it now overlaps the header/existing feed search. Also still a few CSS tweaks/perfections to make.

screen shot 2016-03-13 at 7 48 17 pm

Depends on #3971.
Supersedes #3878.
Closes #943.
Closes #944.

### CODE REVIEW

**Comment by jancavan** [comment]
@TooTallNate exporting works great! But importing doesn't work for me?

No match message also looks like this:
screenshot 2016-03-14 15 44 42

Needs to be (current version on prod):
screenshot 2016-03-14 15 47 40

**Comment by TooTallNate** [comment]
> But importing doesn't work for me?

Does the file picker appear at least? Also check the Network tab in web inspector to see what's going on.

**Comment by jancavan** [comment]
> Does the file picker appear at least?

Yes, it does.

**Comment by jancavan** [comment]
Discussed on Slack, but adding it here for completeness. Nothing happens on the Network tab, but I get this error on the Console tab:

screenshot 2016-03-14 15 55 50

**Comment by TooTallNate** [comment]
Looking into why the Import endpoint gives back a server error:

``` json
{
  "code": 500,
  "headers": [
    {
      "name": "Content-Type",
      "value": "application\/json"
    }
  ],
  "body": {
    "error": "import_job_failed",
    "message": "Unable to queue import job"
  }
}
```

**Comment by jancavan** [comment]
@TooTallNate Nice, importing now works! Not sure if this is the intended behavior, but my imported items only show up after a refresh?

**Comment by TooTallNate** [comment]
> Not sure if this is the intended behavior, but my imported items only show up after a refresh?

At the moment that is intended, until we have a mechanism to detect when the import job has completed. Definitely something I'm planning on figuring out next.

**Comment by jancavan** [comment]
Ah, got it. That's fine. Is that for a different PR? Because until then, I think we can just show a notification that their list has been imported. Asking users to refresh isn't great, but it's better than having no notification at all.

**Importing success notification** (reader.xml is the filename here):
notifications001

**Exporting success notification** (I think for the most part this happens really quickly, but it's still good to inform users if the action they took has been successful or not):
notifications002

We should also display appropriate messaging for when there's problems with the import/export. This is something generic, but what possible error messages occur?
notifications003

For the transactional email, it's telling me I'm already following this blog. It's actually a site that doesn't exist so this is a failed import:
screenshot 2016-03-16 11 41 37

^ Should this also say "sites" instead of "blogs"?

**Comment by blowery** [comment]
Yeah there's a bunch to clean up in that email.

/read/following should just be /
blogs -> sites
all the site URLs should be site names if we have them along with the URL

also nate and I noticed that import was _really_ slow during yesterday's testing. It might be faster now that we're matching properly, but something to keep an eye on while we're testing.

**Comment by TooTallNate** [comment]
> all the site URLs should be site names if we have them along with the URL

The first 2 on that list are easy, but this one would require that we change that email's Content-Type to be HTML, and reformat the message to use HTML rather than plain text (use `<p>`s and `<li>`s, etc.).

> also nate and I noticed that import was really slow during yesterday's testing

If it stays slow, I was thinking it would be neat to make the async import job relay "progress" events for the job and then in the UI we could display "Added '<site name>' to your Reader" for each addition as a (Global) Notice.

**Comment by TooTallNate** [comment]
@jancavan Added the notices; let me know what you think:

screen shot 2016-03-17 at 2 40 02 pm

**Comment by jancavan** [comment]
@TooTallNate The success messages look good :+1: 

The error message though needs to be put more in layman's terms :) What causes the importing problem?

**Comment by TooTallNate** [comment]
> The error message though needs to be put more in layman's terms :) What causes the importing problem?

I'm looking into making the server return more helpful error messages, like "Does not appear to be an OPML file" and other more user-friendly messages.

**Comment by jancavan** [comment]
> "Does not appear to be an OPML file"

We could use "The file format isn't valid. Please see our supported list of file formats." which we should then link to our Support page for OMPL import/export, which I don't think we have yet.

I also think it would be nice to display messages that sound more human and randomized :thought_balloon:

**Comment by jancavan** [comment]
@TooTallNate the notices should also appear in chronological order. The notice for the last performed action should be at the very top. In this example, what I did was:

1. export
2. import
3. import (erroneous file)

But what I'm seeing is this order:
screenshot 2016-03-18 10 50 37

Should be:
screenshot-2016-03-18-10 50 37

Not giving the notices a duration is good for accessibility reasons, but we shouldn't bombard users with notices either :) So in the example above ^, if a user had to export the file again, the first "export success" notice should fade out and appear back up at the very top - essentially showing just one notice for the exact same action that's performed in a series.

**Comment by jancavan** [comment]
@TooTallNate Looking really good so far, there's just a couple items that need to be fixed:

When you aren't following any site, can we just remove the search?
screenshot 2016-03-22 14 50 20

There's also seems to be a bug with the site filter. It doesn't work anymore after refreshing. Steps to reproduce:

1. import a file
2. success notification appears
2. use the site filter
3. hit refresh from notification (can also just be a regular browser refresh) - it also seems like my entire list shows up for a second the moment I hit refresh?
4. user filter again. this works once but after you clear it, it doesn't work anymore

Let's keep it the way it is now where we don't reset filters after a refresh, but it just needs to work again :)

I'll share the video.

**Comment by TooTallNate** [comment]
> When you aren't following any site, can we just remove the search?

Maybe same for the "Export" button or is that overkill?

**Comment by jancavan** [comment]
> Maybe same for the "Export" button or is that overkill?

Oh right, that's a good point :)

**Comment by jancavan** [comment]
@TooTallNate this works great now :+1: 

I don't think we have a great empty state though. I'll re-explore and should probably be a separate PR.

**Comment by TooTallNate** [comment]
Thanks for the review round @jancavan! Merging, and let's follow up with nits in future issues / PRs :grinning:

**Comment by rralian** [comment]
Hey @TooTallNate, can we squash PRs like these before merging? I'm trying to find the root cause of an issue that's happened in the last several days, and it makes it much more difficult when a PR like this gets merged with 25 commits.

**Comment by TooTallNate** [comment]
@rralian :+1: will do, sorry about that
